kotest-common/src/commonMain/kotlin/io/kotest/common/testTimeSource.kt

```diff
@@ -0,0 +1,50 @@
+package io.kotest.common
+
+import kotlin.coroutines.ContinuationInterceptor
+import kotlin.coroutines.coroutineContext
+import kotlin.time.TimeSource
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.test.TestDispatcher
+import kotlinx.coroutines.test.TestCoroutineScheduler
+
+/**
+ * Returns the [TimeSource] used in tests.
+ *
+ * This is [TimeSource.Monotonic] or virtual time, depending on the scheduler in use.
+ */
+@OptIn(KotestInternal::class)
+suspend fun testTimeSource(): TimeSource =
+   coroutineContext.testCoroutineSchedulerOrNull?.timeSource ?: TimeSource.Monotonic
+
+@KotestInternal
+val CoroutineContext.testCoroutineSchedulerOrNull: TestCoroutineScheduler?
+   get() = when (val dispatcher = this[ContinuationInterceptor]) {
+      is TestDispatcher -> dispatcher.scheduler
+      else -> null
+   }
+
+/**
+ * Returns the [TimeSource] used for non-deterministic test functions.
+ *
+ * This is [TimeSource.Monotonic] or – if separately enabled – virtual time, depending on the scheduler in use.
+ */
+@OptIn(KotestInternal::class)
+suspend fun nonDeterministicTestTimeSource(): TimeSource =
+   coroutineContext.nonDeterministicTestCoroutineSchedulerOrNull?.timeSource ?: TimeSource.Monotonic
+
+@OptIn(KotestInternal::class)
+private val CoroutineContext.nonDeterministicTestCoroutineSchedulerOrNull: TestCoroutineScheduler?
+   get() = if (this[NonDeterministicTestVirtualTimeEnabled] != null) testCoroutineSchedulerOrNull else null
```

Comment:
I have a KDoc suggestion too: add a reference to the flag that enables these functions.

```suggestion
/**
 * Returns the [TimeSource] used for non-deterministic test functions.
 *
 * This is [TimeSource.Monotonic] or – if separately enabled – virtual time, depending on the scheduler in use.
 *
 * @see io.kotest.core.spec.Spec.nonDeterministicTestVirtualTimeEnabled
 */
@OptIn(KotestInternal::class)
suspend fun nonDeterministicTestTimeSource(): TimeSource =
   coroutineContext.nonDeterministicTestCoroutineSchedulerOrNull?.timeSource ?: TimeSource.Monotonic

/**
 * Returns the [TimeSource] used for non-deterministic test functions, or `null` if it is not present.
 *
 * @see nonDeterministicTestTimeSource
 * @see io.kotest.core.spec.Spec.nonDeterministicTestVirtualTimeEnabled
 */
@OptIn(KotestInternal::class)
private val CoroutineContext.nonDeterministicTestCoroutineSchedulerOrNull: TestCoroutineScheduler?
   get() = if (this[NonDeterministicTestVirtualTimeEnabled] != null) testCoroutineSchedulerOrNull else null
```

Reply:
(Overlapping Suggestions don't work. Addressing this separately.)

Reply:
The cross-reference to `io.kotest.core.spec.Spec.nonDeterministicTestVirtualTimeEnabled` won't work (different module). Better leave it out since it's all internal anyway and is better detected by using IDE lookup?

Adding `@see nonDeterministicTestTimeSource` to the private `nonDeterministicTestCoroutineSchedulerOrNull` pointing to its adjacent call site seems to distract more than it would help.